### PR TITLE
Fix database errors (#601)

### LINKED ISSUE #601: Fix database errors
Currently, we have two occurrences of `sql: expected x destination arguments in Scan, not y` runtime errors popping up:

- [x] one when querying the CS API's `GET /devices` endpoint
- [ ] one in the syncapi's `rowsToStreamEvents` function

Both should be quite trivial to fix as the former should be pretty straightforward and the latter on require a small bit of refactoring.

### PR DESCRIPTION
**Pull request by DrGlitchMX**
Resolves #601 

Signed-off-by: Maximilian Seifert <max.seifert@drglitch.net>

### CODE REVIEW

**Comment by DrGlitchMX** [comment]
I added a simple docker-compose file (`docker-compose-test.yml`) for running the tests in order to make sure to test against the correct versions of the containers.

I recommend invoking it with
 `docker-compose -f docker/docker-compose-test.yml up --build monolith_test `

**Comment by DrGlitchMX** [comment]
Let me know if this MR is actually useful in its current form :smiley:

**Comment by anoadragon453** [comment]
Could you rebase with latest master? The diff is currently 1 million lines :slightly_smiling_face:

**Comment by DrGlitchMX** [comment]
Just rebased again, so this does not stay behind too much. And I should probably merge instead of actual rebasing, shouldn't I?

**Comment by anoadragon453** [comment]
Ah, yes. That's what I meant but called it a rebase...

Merging forward would indeed be the best approach :)

**Comment by DrGlitchMX** [comment]
Will do so in the future :)

File: docker/docker-compose-test.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+version: "3.4"
+services:
+  monolith_test:
+    container_name: dendrite_monolith_test
+    hostname: monolith
+    entrypoint: ["bash", "./docker/test/monolith.sh"]
```
Comment (anoadragon453) [inline docker/docker-compose-test.yml]: `./build-test-lint.sh` essentially does what `monolith.sh` does.

File: docker/test/monolith.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+rm -rf build
```
Comment (Cnly) [inline docker/test/monolith.sh]: Is this clearing the directory `/path/to/dendrite/build` when the container is started through `docker-compose-test.yml` above? I don't see that directory is used somewhere so this is a bit confusing to me.

File: docker/test/monolith.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+bash ./docker/build.sh
+
+TEST_SUITE=unit-test  scripts/travis-test.sh
```
Comment (Cnly) [inline docker/test/monolith.sh]: Please note that we've switched away from Travis and this script is not there anymore.

**Comment by DrGlitchMX** [comment]
Am working on this again, looks like I need to make the pipeline happy.
As I am a total n00bster to Circle-CI, I wouldn't mind getting pushed into the right direction. If sbd. could tell me how I'd best run the pipeline locally, that would be appreciated, too :)

**Comment by Cnly** [comment]
@DrGlitchMX We run [SyTest](https://github.com/matrix-org/sytest) tests on CircleCI, while stuff like building, unit-testing and linting are run on Buildkite. Right now for this PR it seems just the [linting](https://buildkite.com/matrix-dot-org/dendrite/builds/214#ef360df9-fb2a-4ec8-a38e-6cb23a1e00ed) tools are not happy.

So to run what's run on Buildkite, you basically just need to run [`scripts/build-test-lint.sh`](https://github.com/matrix-org/dendrite/blob/master/scripts/build-test-lint.sh). And if you'd like to run SyTest locally, you may refer to [this](https://github.com/matrix-org/dendrite/blob/master/docs/sytest.md) document.

Please feel free to ask if there's anything unclear.

**Comment by anoadragon453** [comment]
> Am working on this again, looks like I need to make the pipeline happy.
> As I am a total n00bster to Circle-CI, I wouldn't mind getting pushed into the right direction. If sbd. could tell me how I'd best run the pipeline locally, that would be appreciated, too :)

You'll want to clone and install [Sytest](https://github.com/matrix-org/sytest) locally, then follow the [Dendrite-specific setup docs](https://github.com/matrix-org/sytest/blob/develop/docs/dendrite-setup.md). This can also all be done in docker as well, join us in [#dendrite-dev:matrix.org](https://matrix.to/#/!yjVgJUsMLYLAONsoii:jki.re?via=matrix.org&via=privacytools.io&via=ru-matrix.org) if you want help with that.

For testing what Buildkite runs locally, simply run `./scripts/build-test-lint.sh`.

**Comment by Cnly** [comment]
Hmm, sometimes go commands may automatically update the entries in `go.mod` and `go.sum`. If that is the case, please revert the changes.

**Comment by DrGlitchMX** [comment]
Will do so. The new CI / testing infrastructure is a bit hard to grasp for me, I fear. I'll do my best to get this clean.

**Comment by DrGlitchMX** [comment]
Will do so. The new CI / testing infrastructure is a bit hard to grasp for me, I fear. I'll do my best to get this clean.

**Comment by DrGlitchMX** [comment]
Another question on this: am I supposed to squash my commits to sth. more legible or are you fine with the way it is?
If you prefer a cleaner commit history, just give me a shout and I will squash this down to sth. more reasonable.

**Comment by Cnly** [comment]
@DrGlitchMX You don't need to squash it on the PR branch - the commits will be squashed into one when this PR is merged into master. And also, the "newly" passing test isn't actually new; it's just that it's name changed. A merge from master to this PR would fix that.

**Comment by DrGlitchMX** [comment]
@Cnly thank you for the additional information - I hope to have fixed the last quirks with this PR.

**Comment by anoadragon453** [comment]
When testing this on my own Debian Linux machine, I received the following nil pointer dereference:

```
--- FAIL: Test_sanityCheckOutputRoomEvents (0.00s)
    output_room_events_table_test.go:62:
        	Error Trace:	output_room_events_table_test.go:62
        	Error:      	Expected nil, but got: &net.OpError{Op:"dial", Net:"tcp", Source:net.Addr(nil), Addr:(*net.TCPAddr)(0xc00016e7e0), Err:(*os.SyscallError)(0xc000161080)}
        	Test:       	Test_sanityCheckOutputRoomEvents
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7f91c8]

goroutine 34 [running]:
testing.tRunner.func1(0xc00016c200)
	/home/user/goroot/src/testing/testing.go:830 +0x392
panic(0x877aa0, 0xcd7420)
	/home/user/goroot/src/runtime/panic.go:522 +0x1b5
github.com/matrix-org/dendrite/syncapi/storage.Test_sanityCheckOutputRoomEvents(0xc00016c200)
	/home/user/code/dendrite/syncapi/storage/output_room_events_table_test.go:64 +0x98
testing.tRunner(0xc00016c200, 0x917fe8)
	/home/user/goroot/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/home/user/goroot/src/testing/testing.go:916 +0x35a
```

**Comment by DrGlitchMX** [comment]
Okay, this latest failing test leaves me wondering. Any pointers would be appreciated.

**Comment by anoadragon453** [comment]
The CI for dendrite is a little broken at the moment as we've moved things to BuildKite. Bear with us and I'll ping you again once it's back online.

**Comment by DrGlitchMX** [comment]
@anoadragon453 still bearing with you, and happily looking forward to the CI getting fixed. Sorry for the spam. I just want to let you people know that I'll be patient with this.
And I am still motivated to further contribute wherever possible :)

**Comment by anoadragon453** [comment]
@DrGlitchMX Getting there slowly, but it's on the backburner...

**Comment by anoadragon453** [comment]
Ok, CI is back on track!

**Comment by anoadragon453** [comment]
To fix the goimports error, simply run `goimports -w nameoffile.go`.

**Comment by anoadragon453** [comment]
@DrGlitchMX I've fixed the small linting error, however the unit tests are now failing.

**Comment by DrGlitchMX** [comment]
@anoadragon453 by the way - is there some specific reason why the tests are not run automatically on push? I did my best to validate things locally, but, you know, the CI is the single source of truth, after all.

**Comment by anoadragon453** [comment]
> @anoadragon453 by the way - is there some specific reason why the tests are not run automatically on push? I did my best to validate things locally, but, you know, the CI is the single source of truth, after all.

Yeah, it's due to build machines being used across multiple PRs, such that someone can make a PR which disables the machine or does something else to the environment it's running on, and prevent builds from happening on other legitimate PRs.

The solution is just to run the PR content in a sandboxed environment (be that VM or container) and then recreate the environment per PR run. We're actively working to enable this, but it's a big project.

So for now, manual it is :)

**Comment by DrGlitchMX** [comment]
Okay, this thing starts getting old :)
My next attempt to starighten things out is going to be to deactivate the rather clumsy (not integrated into usual test-framework) tests I wrote for this.
I'd happily assume this is okay, as the existing test suite might react to this by letting an existing test pass instead of going red.

I'll probably only start with this by end of this week, so if you have any objections to this approach, please let me know.

Thanks for standing by on this topic :smiley:

**Comment by DrGlitchMX** [comment]
I removed my (superfluous?) tests for now, but I fear the build pipeline is holding a grudge against me.

**Comment by Kegsay** [comment]
Thanks for your work on this. I appreciate your patience and perseverence here! It looks like this got patched in [this commit](https://github.com/matrix-org/dendrite/commit/72565f2eeb8988af90590c90c5c8560186db9d62) so I will be closing this PR.

**Pull request closed without merging**